pkg/ingester/user_tsdb.go

```diff
@@ -95,6 +103,41 @@ func (u *userTSDB) Appender(ctx context.Context) storage.Appender {
 	return u.db.Appender(ctx)
 }
 
+func (u *userTSDB) EphemeralAppender(ctx context.Context) (storage.Appender, error) {
+	es := u.getEphemeralStorage()
+	if es != nil {
+		return es.Appender(ctx), nil
+	}
+
+	es, err := u.createEphemeralStorage()
+	if err != nil {
+		return nil, err
+	}
+
+	return es.Appender(ctx), nil
+}
+
+func (u *userTSDB) createEphemeralStorage() (*tsdb.Head, error) {
+	u.ephemeralMtx.Lock()
+	defer u.ephemeralMtx.Unlock()
+
+	if u.ephemeralStorage != nil {
+		return u.ephemeralStorage, nil
+	}
+
+	es, err := u.ephemeralFactory()
+	u.ephemeralStorage = es
```

Comment:
i think it might be better to not assign that if `err != nil`. Otherwise, if in the future somebody changes the implementation of the ephemeral factory, we might suddenly end up in a situation where `u.ephemeralFactory()` returns a non-nil `es` and also an `err` and then we have an erronous value assigned to `u.ephemeralStorage`.

Reply:
Such change would hopefully not pass the review, but there is no harm in doing what you suggest.